fix bug using params from --config file

cdetrio opened the pull request:
Without this, a `byzantineForkBlock` param passed in a json file to `--config` will be overridden by this line in [mainNetwork.cpp](https://github.com/ethereum/cpp-ethereum/blob/648b361f5c65372aac698e27512ccdc4a440f594/libethashseal/genesis/mainNetwork.cpp#L30).

Also, the `ChainParams` initialization was causing seemingly pointless loops over the all the mainnet genesis accounts in `jsonToAccountMap`.

winsvega commented on test/tools/jsontests/vm.cpp:
```diff
@@ -482,8 +482,8 @@ class VmTestFixture
 		}
-		VmTestSuite suite;
-		suite.runAllTestsInFolder(casename);
+		//VmTestSuite suite;
```
whats happened here ?

cdetrio replied:
Fixes this error I get when building:

```
ubuntu@certification-hive-test03:~/cdetrio/cpp-ethereum/build$ make
[  3%] Built target secp256k1
[  6%] Built target cryptopp
[ 10%] Built target jsonrpccpp
[ 13%] Built target mpir
[ 16%] Built target snark
[ 16%] Built target BuildInfo.h
Scanning dependencies of target devcore
[ 16%] Building CXX object libdevcore/CMakeFiles/devcore.dir/Common.cpp.o
[ 17%] Linking CXX static library libdevcore.a
[ 23%] Built target devcore
[ 27%] Built target scrypt
[ 29%] Built target devcrypto
[ 35%] Built target p2p
[ 37%] Built target ethash
[ 41%] Built target ethcore
[ 44%] Built target evm
[ 55%] Built target ethereum
Scanning dependencies of target ethashseal
[ 55%] Building CXX object libethashseal/CMakeFiles/ethashseal.dir/GenesisInfo.cpp.o
[ 55%] Linking CXX static library libethashseal.a
[ 57%] Built target ethashseal
[ 60%] Built target whisper
Scanning dependencies of target webthree
[ 60%] Building CXX object libwebthree/CMakeFiles/webthree.dir/WebThree.cpp.o
[ 61%] Linking CXX static library libwebthree.a
[ 61%] Built target webthree
[ 69%] Built target web3jsonrpc
Scanning dependencies of target eth
[ 69%] Building CXX object eth/CMakeFiles/eth.dir/main.cpp.o
[ 70%] Linking CXX executable eth
[ 71%] Built target eth
Scanning dependencies of target ethkey
[ 72%] Building CXX object ethkey/CMakeFiles/ethkey.dir/main.cpp.o
[ 72%] Linking CXX executable ethkey
[ 72%] Built target ethkey
[ 72%] Linking CXX executable ethvm
[ 73%] Built target ethvm
[ 74%] Linking CXX executable rlp
[ 74%] Built target rlp
Scanning dependencies of target testeth
[ 74%] Building CXX object test/CMakeFiles/testeth.dir/tools/libtesteth/boostTest.cpp.o
[ 74%] Building CXX object test/CMakeFiles/testeth.dir/tools/libtesteth/Stats.cpp.o
[ 75%] Building CXX object test/CMakeFiles/testeth.dir/tools/libtesteth/TestHelper.cpp.o
[ 75%] Building CXX object test/CMakeFiles/testeth.dir/tools/libtesteth/Options.cpp.o
[ 75%] Building CXX object test/CMakeFiles/testeth.dir/tools/libtesteth/ImportTest.cpp.o
[ 76%] Building CXX object test/CMakeFiles/testeth.dir/tools/libtesteth/BlockChainHelper.cpp.o
[ 76%] Building CXX object test/CMakeFiles/testeth.dir/tools/libtesteth/FillJsonFunctions.cpp.o
[ 76%] Building CXX object test/CMakeFiles/testeth.dir/tools/jsontests/vm.cpp.o
[ 76%] Building CXX object test/CMakeFiles/testeth.dir/tools/jsontests/BlockChainTestsBoost.cpp.o
[ 77%] Building CXX object test/CMakeFiles/testeth.dir/tools/jsontests/StateTests.cpp.o
[ 77%] Building CXX object test/CMakeFiles/testeth.dir/tools/jsontests/TransactionTests.cpp.o
[ 78%] Building CXX object test/CMakeFiles/testeth.dir/tools/jsontests/BlockChainTests.cpp.o
[ 78%] Building CXX object test/CMakeFiles/testeth.dir/tools/fuzzTesting/fuzzHelper.cpp.o
[ 78%] Building CXX object test/CMakeFiles/testeth.dir/tools/fuzzTesting/createRandomTest.cpp.o
[ 78%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libdevcore/rlp.cpp.o
[ 79%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libtesteth/blockchainTest.cpp.o
[ 79%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethcore/difficulty.cpp.o
[ 79%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethcore/dagger.cpp.o
[ 79%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethcore/PrecompiledTest.cpp.o
[ 79%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethcore/SealEngineTest.cpp.o
[ 80%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethcore/keymanager.cpp.o
[ 80%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethereum/TransactionQueue.cpp.o
[ 80%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethereum/StateUnitTests.cpp.o
[ 81%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethereum/BlockChainInsert.cpp.o
[ 81%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethereum/EthereumPeerTest.cpp.o
[ 82%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethereum/ExtVMTest.cpp.o
[ 82%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethereum/BlockQueue.cpp.o
[ 83%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethereum/BlockChain.cpp.o
[ 83%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethereum/Transaction.cpp.o
[ 84%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethereum/Genesis.cpp.o
[ 84%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethereum/Block.cpp.o
[ 84%] Building CXX object test/CMakeFiles/testeth.dir/unittests/libethereum/GasPricer.cpp.o
[ 85%] Building CXX object test/CMakeFiles/testeth.dir/unittests/external-dependencies/boost.cpp.o
[ 85%] Linking CXX executable testeth
/home/ubuntu/cdetrio/cpp-ethereum/test/tools/jsontests/vm.cpp:486: error: undefined reference to 'dev::test::TestSute::runAllTestsInFolder(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
collect2: error: ld returned 1 exit status
test/CMakeFiles/testeth.dir/build.make:1787: recipe for target 'test/testeth' failed
make[2]: *** [test/testeth] Error 1
CMakeFiles/Makefile2:1299: recipe for target 'test/CMakeFiles/testeth.dir/all' failed
make[1]: *** [test/CMakeFiles/testeth.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
```

gumb0 commented on eth/main.cpp:
```diff
@@ -327,7 +327,7 @@ int main(int argc, char** argv)
 
 	string jsonAdmin;
-	ChainParams chainParams(genesisInfo(eth::Network::MainNetwork), genesisStateRoot(eth::Network::MainNetwork));
+	ChainParams chainParams;
```
This doesn't seem right. Which chain params & genesis now `eth` without options starts with? Starting `eth` should have the same effect as `eth --mainnet`

cdetrio replied:
Ah, then the default (with no `--config` and no `--mainnet`) should be handled _after_ the `--mainnet` and `--testnet` cases https://github.com/ethereum/cpp-ethereum/blob/908d9f5b89d3fdbf421522d68bd03130ed6f4ba8/eth/main.cpp#L598-L601.  Otherwise the mainnet genesis accounts are needlessly loaded (and seem to interfere with the `--config` params).

cdetrio commented:
@gumb0 I pushed a commit to make`eth` default to mainnet config.

gumb0 commented on eth/main.cpp:
```diff
@@ -595,10 +596,14 @@ int main(int argc, char** argv)
 		else if (arg == "--gas-floor" && i + 1 < argc)
 			gasFloor = u256(argv[++i]);
-		else if (arg == "--mainnet")
+		else if (arg == "--mainnet") {
```
Put a brace on a separate line please

gumb0 commented on eth/main.cpp:
```diff
@@ -595,10 +596,14 @@ int main(int argc, char** argv)
-		else if (arg == "--ropsten" || arg == "--testnet")
+			chainConfigIsSet = true;
+		}
+		else if (arg == "--ropsten" || arg == "--testnet") {
```
Here, too

gumb0 commented on eth/main.cpp:
```diff
@@ -841,6 +848,10 @@ int main(int argc, char** argv)
 
+	if (!chainConfigIsSet)
+		// default to mainnet if not already set with any of `--mainnet`, `--testnet`, `--genesis`, `--config`
+		ChainParams chainParams(genesisInfo(eth::Network::MainNetwork), genesisStateRoot(eth::Network::MainNetwork));
```
This won't work, it creates another `chainParams` variable which exists only in the scope of this `if` (so destroyed immediately).

I suggest declaring it as `boost::optional<ChainParams> chainParams;` then you don't need the flag, assign to it as usual `chainParams = ChainParams(...)`, test if it's set with `if (chainParams)` and pass it further as `*chainParams`

cdetrio commented:
@gumb0  addressed your comments the lazy way for now.

I tried doing it the other way with `boost::optional`, but it raised further complications:  https://github.com/ethereum/cpp-ethereum/compare/develop...cdetrio:chainparams-option-attempt?expand=1

```
[ 69%] Building CXX object eth/CMakeFiles/eth.dir/main.cpp.o
/home/ubuntu/cdetrio/cpp-ethereum/eth/main.cpp: In function ‘int main(int, char**)’:
/home/ubuntu/cdetrio/cpp-ethereum/eth/main.cpp:799:30: error: ‘class boost::optional<dev::eth::ChainParams>’ has no member named ‘loadConfig’
    chainParams = chainParams.loadConfig(configJSON);
                              ^
/home/ubuntu/cdetrio/cpp-ethereum/eth/main.cpp:815:30: error: ‘class boost::optional<dev::eth::ChainParams>’ has no member named ‘loadGenesis’
    chainParams = chainParams.loadGenesis(genesisJSON);
                              ^
/home/ubuntu/cdetrio/cpp-ethereum/eth/main.cpp:843:15: error: ‘class boost::optional<dev::eth::ChainParams>’ has no member named ‘extraData’
   chainParams.extraData = sha3(privateChain).asBytes();
               ^
/home/ubuntu/cdetrio/cpp-ethereum/eth/main.cpp:844:15: error: ‘class boost::optional<dev::eth::ChainParams>’ has no member named ‘difficulty’
   chainParams.difficulty = chainParams.minimumDifficulty;
               ^
/home/ubuntu/cdetrio/cpp-ethereum/eth/main.cpp:844:40: error: ‘class boost::optional<dev::eth::ChainParams>’ has no member named ‘minimumDifficulty’
   chainParams.difficulty = chainParams.minimumDifficulty;
                                        ^
/home/ubuntu/cdetrio/cpp-ethereum/eth/main.cpp:845:15: error: ‘class boost::optional<dev::eth::ChainParams>’ has no member named ‘gasLimit’
   chainParams.gasLimit = u256(1) << 32;
               ^
/home/ubuntu/cdetrio/cpp-ethereum/eth/main.cpp:915:15: error: ‘class boost::optional<dev::eth::ChainParams>’ has no member named ‘sealEngineName’
   chainParams.sealEngineName = "NoProof";
               ^
/home/ubuntu/cdetrio/cpp-ethereum/eth/main.cpp:916:15: error: ‘class boost::optional<dev::eth::ChainParams>’ has no member named ‘allowFutureBlocks’
   chainParams.allowFutureBlocks = true;
```

gumb0 commented:
Yeah that would be fixed with using `->` instead of `.`, bit ok, your version will do, too
